### PR TITLE
RLC: permit ownership transfer during field re-assignment

### LINKED ISSUE #5971: Resource Leak Checker: false positive from ownership transfer during re-assignment
This is a sub-issue of #5969 (where it is described as false positive 1). Consider the following code, which I've checked into the `cm-ownership-transfer-at-reassignment` branch of my fork:
```
import org.checkerframework.checker.calledmethods.qual.EnsuresCalledMethods;
import org.checkerframework.checker.mustcall.qual.CreatesMustCallFor;
import org.checkerframework.checker.mustcall.qual.InheritableMustCall;
import org.checkerframework.checker.mustcall.qual.Owning;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.io.IOException;

@InheritableMustCall("disconnect")
public class OwnershipTransferAtReassignment {

  private @Owning @Nullable Node head = null;

  @CreatesMustCallFor("this")
  public boolean add() {
    head = new Node(head);
    return true;
  }

  @EnsuresCalledMethods(value="this.head", methods="disconnect")
  public void disconnect() {
    head.disconnect();
  }

  @InheritableMustCall("disconnect")
  private static class Node {
    @Owning private final @Nullable Node next;

    public Node(@Owning @Nullable Node next) {
      this.next = next;
    }

    @EnsuresCalledMethods(value="this.next", methods="disconnect")
    public void disconnect() {
      next.disconnect();
    }
  }
}
```

Currently, the checker issues a warning when `head` is re-assigned at line 16. No warnings should be issued (the re-assignment is safe, because ownership is transferred during the re-assignment to the `Node` constructor).

### PR DESCRIPTION
**Pull request by kelloggm**
Fixes #5971

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java [hunk 1]
```diff
@@ -1385,6 +1385,41 @@ && varTrackedInObligations(obligations, (LocalVariableNode) receiver))
       checkEnclosingMethodIsCreatesMustCallFor(node, enclosingMethodTree);
     }
 
+    // Special case: re-assigning a field to a value in an owning position on the RHS is
+    // permitted. For example, if the field is a class whose constructor takes another instance
+    // of itself (such as a node in a linked list) in an owning position, re-assigning the field
+    // to a new instance that takes the field's value as an owning parameter is safe (the new value
+    // has taken responsibility for closing the old value). In that case, it is safe to skip the
+    // must call vs called methods check below (but an @CreatesMustCallFor annotation is still
+    // required, so this special case must be here).
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1394]: I find this comment to be a bit hard to parse.  Let me write what I think is going on and then you can decide how to edit / integrate.  The special case we are handling here is where the field being assigned is itself getting passed in an owning position to another method on the RHS of the assignment, e.g.:
```java
Foo takesOwnership(@Owning Foo foo) { ... }
...
@CreatesMustCallFor("this")
void overwrite() {
  this.f = takesOwnership(f);
} 
```
In such a case, it is not required that the must-call obligation of the field be satisfied via method calls before the assignment, since the invoked method will take ownership of the object previously referenced by the field and handle the obligation.

Taking a step back, I wish we didn't need yet another special case here.  But, I can't immediately think of a significantly cleaner solution, so we can go with this for now.

**Review by msridhar — CHANGES_REQUESTED** [review]
Code change looks good but I think the docs could be tweaked a bit

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java [hunk 2]
```diff
@@ -1385,6 +1385,41 @@ && varTrackedInObligations(obligations, (LocalVariableNode) receiver))
...
+    Node rhs = node.getExpression();
+    if (!noLightweightOwnership
```
[the remark this answers is not in the document]
Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1396]: I don't have a strong feeling about this. I don't mind maintaining them (and keeping them means it'll be easy to re-run those experiments with a new RLC version, if we ever want to).

I'd approve a PR to remove them, but I don't think I'll make the PR myself.

**Review by msridhar — DISMISSED** [review]

**Comment by kelloggm** [comment]
@msridhar there was a merge conflict, so this needs to be re-approved in order to merge

**Review by msridhar — APPROVED** [review]
